### PR TITLE
Implement missing implementations of toString(), fixing an issue mentioned in #1196. Also optimize some implementations of toString() that were delegating to iterators.

### PR DESCRIPTION
**Pull request by motlin**

### CODE REVIEW

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 1]
```diff
@@ -847,7 +847,7 @@ public void clear()
         @Override
         public String toString()
         {
-            return Iterate.makeString(this, "[", ", ", "]");
+            return AbstractMutableBiMap.this.delegate.keySet().toString();
         }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java:851]: keySet().toString() was delegating to an iterator

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 2]
```diff
@@ -965,11 +965,11 @@ public void clear()
         @Override
         public String toString()
         {
-            return Iterate.makeString(this, "[", ", ", "]");
+            return AbstractMutableBiMap.this.delegate.values().toString();
         }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java:969]: values().toString() was delegating to an iterator

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java [hunk 3]
```diff
@@ -991,6 +991,12 @@ public int hashCode()
             return AbstractMutableBiMap.this.hashCode();
         }
 
+        @Override
+        public String toString()
+        {
+            return AbstractMutableBiMap.this.delegate.entrySet().toString();
+        }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/AbstractMutableBiMap.java:998]: entrySet().toString() wasn't implemented and was printing the wrong string

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java [hunk 4]
```diff
@@ -269,18 +267,6 @@ public V forcePut(K key, V value)
...
-    {
-        return LazyIterate.adapt(this.keySet());
-    }
-
-    @Override
-    public RichIterable<V> valuesView()
-    {
-        return LazyIterate.adapt(this.values());
-    }
-
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/bimap/mutable/SynchronizedBiMap.java:283]: It's odd that the implementation of keysView, valuesView, and keyValuesView got split up in some cases. These were moved up to the superclass and all the implementations are together now.

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/AbstractSynchronizedMapIterable.java [hunk 5]
```diff
@@ -350,15 +349,22 @@ public V updateValue(K key, Function0<? extends V> factory, Function<? super V,
...
     @Override
     public RichIterable<Pair<K, V>> keyValuesView()
     {
-        synchronized (this.lock)
-        {
-            Set<Entry<K, V>> entries = this.getDelegate().entrySet();
-            Iterable<Pair<K, V>> pairs = Iterate.collect(entries, AbstractImmutableEntry.getPairFunction());
-            return LazyIterate.adapt(pairs);
-        }
+        return LazyIterate.adapt(this.entrySet()).collect(AbstractImmutableEntry.getPairFunction());
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/map/AbstractSynchronizedMapIterable.java:367]: This implementation was wrong. By delegating to a non-lazy implementation of collect(), it caused observable changes in iteration order on collections that have a sort order using a comparator.

The test MapIterableTestCase#Iterable_toString() catches the bug when run in SynchronizedSortedMapTest. The assertion:

```java
Assert.assertEquals(
        "[10:4, 9:4, 8:4, 7:4, 6:3, 5:3, 4:3, 3:2, 2:2, 1:1]",
        this.newWith(4, 4, 4, 4, 3, 3, 3, 2, 2, 1).keyValuesView().toString());
```

Fails with:

```
Expected :[10:4, 9:4, 8:4, 7:4, 6:3, 5:3, 4:3, 3:2, 2:2, 1:1]
Actual   :[4:3, 7:4, 8:4, 10:4, 9:4, 6:3, 5:3, 3:2, 2:2, 1:1]
```

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/UnifiedMap.java [hunk 6]
```diff
@@ -3228,7 +3245,12 @@ private void chainedAddToList(Object[] chain, FastList<V> replace)
...
         public String toString()
         {
-            return Iterate.makeString(this, "[", ", ", "]");
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append('[');
+            Procedure<V> appendStringProcedure = new AppendStringProcedure<>(stringBuilder, ", ");
+            this.forEach(appendStringProcedure);
+            stringBuilder.append(']');
+            return stringBuilder.toString();
         }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/map/mutable/UnifiedMap.java:3254]: All three implementations of toString() were delegating to an iterator

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/map/strategy/mutable/UnifiedMapWithHashingStrategy.java [hunk 7]
```diff
@@ -2852,7 +2869,12 @@ private void chainedAddToList(Object[] chain, FastList<V> replace)
...
         public String toString()
         {
-            return Iterate.makeString(this, "[", ", ", "]");
+            StringBuilder stringBuilder = new StringBuilder();
+            stringBuilder.append('[');
+            Procedure<V> appendStringProcedure = new AppendStringProcedure<>(stringBuilder, ", ");
+            this.forEach(appendStringProcedure);
+            stringBuilder.append(']');
+            return stringBuilder.toString();
         }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/map/strategy/mutable/UnifiedMapWithHashingStrategy.java:2878]: All three implementations of toString() were delegating to an iterator

File: eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/Iterate.java [hunk 8]
```diff
@@ -130,6 +131,10 @@ private Iterate()
         {
             ((InternalIterable<T>) iterable).forEach(procedure);
         }
+        else if (iterable instanceof BatchIterable)
+        {
+            ((BatchIterable<T>) iterable).forEach(procedure);
+        }
```
Comment (motlin) [inline eclipse-collections/src/main/java/org/eclipse/collections/impl/utility/Iterate.java:137]: We hit this path when running `map.keysView().toString()` since:

- keysView() is implemented by returning `LazyIterate.adapt(this.keySet())`
- the `LazyIterableAdapter` eventually delegates to `Iterate.forEach()`
- the keys/values/entries inner classes in UnifiedMap are BatchIterables, not RichIterables

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java [hunk 9]
```diff
@@ -11,17 +11,101 @@
...
 import org.eclipse.collections.impl.factory.Lists;
 import org.eclipse.collections.test.RichIterableUniqueTestCase;
+import org.eclipse.collections.test.map.MapIterableTestCase;
+import org.junit.Assert;
+import org.junit.Test;
 
 import static org.eclipse.collections.test.IterableTestCase.assertEquals;
 
-public interface BiMapTestCase extends RichIterableUniqueTestCase
+public interface BiMapTestCase extends RichIterableUniqueTestCase, MapIterableTestCase
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java:24]: BiMaps are MapIterables, so I added the interface here. This causes more diamonds so more methods need to be overridden. In addition, some map tests rely on data where 2 different keys store the same value. This isn't allowed in BiMaps so a few tests are overridden. These assertions could be made even more general with fewer overrides, but this code review is already a lot bigger than I was expecting.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java [hunk 10]
```diff
@@ -11,17 +11,101 @@
...
+        assertEquals(this.newMutableForFilter("3Three", "2Two", "1One"), forEachKeyValue2);
+
+        MutableCollection<Integer> forEachValue = this.newMutableForFilter();
+        bimap.forEachValue(value -> forEachValue.add(value + 10));
+        assertEquals(this.newMutableForFilter(13, 13, 13, 12, 12, 11), forEachValue);
+
+        MutableCollection<Object> forEachKey = this.newMutableForFilter();
+        bimap2.forEachKey(key -> forEachKey.add(key + 1));
+        assertEquals(this.newMutableForFilter(4, 3, 2), forEachKey);
+    }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java:90]: This override is because the superclass uses duplicate values in its test data.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java [hunk 11]
```diff
@@ -11,17 +11,101 @@
...
+    @Override
+    default void MapIterable_flipUniqueValues()
+    {
+        BiMap<String, Integer> bimap = this.newWithKeysValues("Three", 3, "Two", 2, "One", 1);
+        BiMap<Integer, String> result = bimap.flipUniqueValues();
+
+        assertEquals(
+                this.newWithKeysValues(3, "Three", 2, "Two", 1, "One"),
+                result);
+    }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/BiMapTestCase.java:101]: This override is because the superclass tests that an assertion is thrown when flipping a map with duplicate values. A BiMap cannot have duplicate values and never throws from flipUniqueValues().

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/UnsortedBiMapTestCase.java [hunk 12]
```diff
@@ -49,23 +57,4 @@ default void Iterable_remove()
...
-        Matcher matcher = pattern.matcher(string);
-        Assert.assertTrue(string, matcher.matches());
-
-        assertEquals(
-                Bags.immutable.with("1", "2", "3"),
-                Bags.immutable.with(
-                        matcher.group(2),
-                        matcher.group(4),
-                        matcher.group(6)));
-    }
```
Comment (motlin) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/UnsortedBiMapTestCase.java:70]: This was a lot of effort to make assertions about the values when the map contains random keys. The new test is just as good for toString() and covers the views as well.

File: checkstyle-configuration.xml [hunk 13]
```diff
@@ -347,7 +347,7 @@
 
         <module name="AvoidStaticImport">
             <property name="excludes"
-                value="org.junit.Assert.*,org.junit.Assume.*,org.hamcrest.CoreMatchers.*,org.hamcrest.Matchers.*,org.eclipse.collections.impl.test.Verify.*,org.eclipse.collections.impl.factory.Iterables.*,org.eclipse.collections.test.IterableTestCase.*" />
+                value="org.junit.Assert.*,org.junit.Assume.*,org.hamcrest.CoreMatchers.*,org.hamcrest.Matchers.*,org.hamcrest.MatcherAssert.*,org.eclipse.collections.impl.test.Verify.*,org.eclipse.collections.impl.factory.Iterables.*,org.eclipse.collections.test.IterableTestCase.*" />
```
Comment (motlin) [inline checkstyle-configuration.xml:350]: Junit's Assert.assertThat() is deprecated and says `@deprecated use {@code org.hamcrest.MatcherAssert.assertThat()}`

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/immutable/ImmutableHashBiMapInverseTest.java [hunk 14]
```diff
@@ -30,10 +33,23 @@
...
-            {
-                throw new IllegalStateException();
-            }
+            assertNull(result.put(each, random.nextDouble()));
+        }
+        return result.toImmutable().inverse();
+    }
+
+    @Override
+    public <K, V> ImmutableBiMap<K, V> newWithKeysValues(Object... elements)
```
Comment (prathasirisha) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/immutable/ImmutableHashBiMapInverseTest.java:42]: How is this related to the toString() method in 1196?

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java [hunk 15]
```diff
@@ -42,6 +48,33 @@
...
+        assertEquals(Optional.of(Integer.valueOf(1)), ((OrderedIterable<?>) this.newWith(3, 3, 3, 2, 2, 1)).getFirstOptional());
+    }
+
+    @Override
+    @Test
+    default void OrderedIterable_getLastOptional()
+    {
+        assertEquals(Optional.of(Integer.valueOf(3)), ((OrderedIterable<?>) this.newWith(3, 3, 3, 2, 2, 1)).getLastOptional());
+    }
+
```
[the remark this answers is not in the document]
Comment (prathasirisha) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java:77]: okay - but I am confused as to why this is in this PR 😕 How is this related to toString() issue from 1196.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/mutable/HashBiMapNoIteratorTest.java [hunk 16]
```diff
@@ -0,0 +1,144 @@
+/*
+ * Copyright (c) 2022 Goldman Sachs and others.
```
Comment (prathasirisha) [inline unit-tests-java8/src/test/java/org/eclipse/collections/test/bimap/mutable/HashBiMapNoIteratorTest.java:2]: Company info needs to be updated.

File: unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java (hunk 15 above)
[the remark this answers is not in the document]
Comment (motlin) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java:77]: To be honest, I don't remember how I came across this anymore, because the bugs I found were mostly in Maps, not Bags. However, I wound up stepping through a lot of tests in the debugger to find and fix the bugs, so I probably wound up stepping through a subclass of this TestCase. Once I saw it, it was obviously wrong to me because no assertions should ever be in unit-tests-java8/**/*Test.java. Assertions only go in *TestCase.java.

[the remark this answers is not in the document]
Comment (motlin) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java:77]: Now that I'm looking around, there are other minor violations of this rule. Specifically, the subclasses of MultiReaderMutableCollectionTestCase are making assertions about classes instead of interfaces. I'm not planning on fixing the problem in general. I just recognized this mistake as one that I wrote so I wanted to fix it myself.

[the remark this answers is not in the document]
Comment (prathasirisha) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java:77]: Understood. 

Can you split these into separate commits for ease of review and for managing release notes? Thank you!

[the remark this answers is not in the document]
Comment (motlin) [reply unit-tests-java8/src/test/java/org/eclipse/collections/test/bag/mutable/sorted/MutableSortedBagNoComparatorTestCase.java]: I pushed the tests back down to the subclass / reverted this change. And I rebased on upstream/master.

**Comment by motlin** [comment]
I keep rebasing this. It addresses real bugs. Could we get it merged?

**Review by prathasirisha — APPROVED** [review]

**Comment by donraab** [comment]
@motlin please rebase one last time and I will merge now that @prathasirisha has marked her review as approved.

**Comment by motlin** [comment]
@donraab rebased!
